Add tests CU-861mt6hkx

big-nacho opened the pull request:
This PR adds test functionality to vectorizer

To test, run `python -m pytest vectorizing/tests/test.py`

Testing pipeline is as follows:
1. A config object in `config.py` is used to run tests in combination with the images placed in `images`
2. Since the server needs a URL per request, we generate them by uploading to an S3 bucket
3. Each image is:
    - Vectorized, once per test case
    - Rasterized to `png` via `libcairo`
    - If it's the first time the test is run, the rasterized version is placed in `results`, otherwise, the test result is compared to the already stored one via `uqi`. See below what happens when the comparison fails.

If a comparison fails (the test result is too far apart from the expected result), an entry is placed in `diff_output` that highlights parts where the images differ too much. To test this you can just run tests and then replace any of the images in result with some other one (keeping the same name of course)

<details>

<summary>Example</summary>

![shapes_2](https://github.com/Kittl/vectorizing/assets/67847653/a13b68c8-31a5-4749-ba65-9bfc4e1ead0d)


</details>

big-nacho commented on requirements/dev.in:
```diff
@@ -2,7 +2,9 @@
\ No newline at end of file
+pytest
+cairosvg
+sewar
```
Should we add versions to `cairosvg` and `sewar`?

pietrobolcato replied:
Yes please :)

When inside the conda dev env you can find them with pip show:

```
conda activate dev
pip show cairosvg
pip show sewar
```

big-nacho replied:
Added versions :)

big-nacho commented on vectorizing/util/limit_size.py:
```diff
@@ -2,7 +2,7 @@
 
 # Size restrictions for image
-MIN_PIXEL_COUNT = 64
+MIN_PIXEL_COUNT = 8
```
This is a random change, I think `64` was a bit too high

big-nacho reviewed (commented):
@pietrobolcato Tests are a bit random atm, excluding white text. I think another relevant one is transparent background + color foreground + color solver, but I'd rather add that while fixing a small bug we have for those cases. 0px * 0px I also thought about, but after a bit of research I found out most image formats specifications deem 0 invalid for any image dimension.

pietrobolcato commented on vectorizing/tests/conftest.py:
```diff
@@ -0,0 +1,12 @@
+from vectorizing import create_app
+
+@pytest.fixture()
+def app():
```
is this fixture being used in any place, or it has another function? or can we perhaps have a single one, eg:

```
@pytest.fixture()
def client(app):
    app = create_app()
    app.config.update({ "TESTING": True })
    return app.test_client()
```

pietrobolcato commented on docker-compose.yml:
We should update the part about tests in docker compose as well, the current command doesn't work:

`pytest /app/src/test/test.py`

big-nacho replied:
deleted

pietrobolcato reviewed (changes requested):
I coulnd't run the tests, I didn't remember the test bucket name precisely and it was not set in .env.example - if you update it, i'll try again :) 

But in general good stuff dear Nacho!

pietrobolcato commented:
> I think another relevant one is transparent background + color foreground + color solver, but I'd rather add that while fixing a small bug we have for those cases

I think adding this test and solving the small bug you have already in mind could be actually a really good addition to this PR instead of waiting for another one, given that we had this problem in prod already and would be great to make sure it doesn't happen again - what do you think? :)

big-nacho commented:
> Few notes:
> 
> * I coulnd't run the tests, I didn't remember the test bucket name precisely and it was not set in .env.example - if you update it, i'll try again :)
> * > I think another relevant one is transparent background + color foreground + color solver, but I'd rather add that while fixing a small bug we have for those cases
>   
>   
>   I think adding this test and solving the small bug you have already in mind could be actually a really good addition to this PR instead of waiting for another one, given that we had this problem in prod already and would be great to make sure it doesn't happen again - what do you think?
> * We should also think about setting up a minimal CI pipeline - If you like I can definitely help there, let's speak about it tomorrow
> 
> In general good stuff dear Nacho! Thanks for the good addition!

I've addressed issues, as well as added comments for functions to make things clearer. I also added a quick separate metric for images that are too small. UQI seems to fail with nan for those, not sure why, but I don't think it's too relevant. I've rolled back the minimum pixel size as well, because I hadn't realized, but `faiss` gives a warning of few data points. I think it's best to avoid it.

Regarding the raised concerns:
1. I don't think we should have the bucket name available. Since the repo is public, I would hide as much info as possible.
2. We can merge everything together, but I would open a new PR for that bug, to separate things clearly
3. I would very much appreciate help with the CI setup, feel free to commit here too :)
